#128 Distribution Model in Admin Panel and #129 API

katalyzator opened the pull request:
- Added Distribution model
- Added new model to Admin Panel with some search and filter configurations
- Added constants file to separate const variables

katalyzator committed:
Merge branch 'master' of https://github.com/Entea/covid-supply-info

katalyzator committed:
Merge branch 'master' of https://github.com/Entea/covid-supply-info

katalyzator committed:
Merge branch 'master' of https://github.com/Entea/covid-supply-info

katalyzator committed:
Merge branch 'master' of https://github.com/Entea/covid-supply-info

katalyzator committed:
Merge branch 'master' of https://github.com/Entea/covid-supply-info

katalyzator committed:
Merge branch 'master' of https://github.com/Entea/covid-supply-info

katalyzator committed:
Merge branch 'master' of https://github.com/Entea/covid-supply-info

katalyzator committed:
Merge branch 'master' of https://github.com/Entea/covid-supply-info

katalyzator committed:
added new Distribution model

katalyzator committed:
added new Distribution model

katalyzator committed:
implemented API part

katalyzator committed:
implemented API part

isabek commented on backend/distributor/filters.py:
```diff
@@ -0,0 +1,24 @@
+
+        self.lookup_expr = 'in'
+        values = value.split(',')
+        return super(MultipleListFilter, self).filter(qs, values).distinct()
```
Let's set some limit.

katalyzator replied:
What limit should I put there? Maybe `10`?

isabek replied:
20 is okay.

isabek commented on backend/distributor/models.py:
```diff
@@ -340,3 +342,23 @@ class Meta:
+    receiver = models.TextField(verbose_name='Кто принял?', blank=True)
+    date_of_distribute = models.DateField(verbose_name='Дата распределения')
+    status = models.CharField(max_length=20, choices=DISTRIBUTION_STATUSES, verbose_name='Статус')
+    donations = models.ManyToManyField(Donation, verbose_name='Пожертвования')
```
```suggestion
    donations = models.ManyToManyField(Donation, verbose_name=_('Пожертвования'))
```

isabek replied:
+ надо добавить дату отправки и дату доставки. 
+ по умолчанию должен быть статус READY_TO_SEND (@Entea @BupycHuk ).

katalyzator replied:
@Isabek `date_of_distribute` вроде это и есть дата отправки и вот дату доставки я добавлю

isabek replied:
@katalyzator может быть следовать одному формату

created_at
distributed_at

isabek replied:
+ take into account that fact we will add a multilingual feature in the future.

katalyzator committed:
added default value and refactored

katalyzator committed:
added default value and refactored

katalyzator committed:
changed field names

katalyzator committed:
changed field names

isabek commented on backend/distributor/constants.py:
```diff
@@ -0,0 +1,9 @@
+DELIVERED = 'delivered'
+
+DISTRIBUTION_STATUSES = (
+    (READY_TO_SEND, 'Подготовлено'),
```
please, wrap status texts into `_()_`

katalyzator committed:
wrapped statuses text

katalyzator committed:
wrapped statuses text

BupycHuk commented on backend/distributor/api.py:
```diff
@@ -168,13 +169,26 @@ class ContactMessageAPIView(APIView):
-            )
-            return Response({'success': True}, status=status.HTTP_201_CREATED)
-        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+        serializer.is_valid(raise_exception=True)
```
Why do we raise exception, instead of returning normal response?

katalyzator replied:
It already returns normal response, I just reduce code lines, nothing has changed, it remains as before. Instead of checking inside if statement we can set `raise_exception=True` and it will return `Bad Request` with `status code 400` and response body will be with detailed error messages

BupycHuk commented:
do we need detailed error response in public API? I don't think so

ср, 8 апр. 2020 г., 16:52 Aibek Raiymbekov <notifications@github.com>:

> *@katalyzator* commented on this pull request.
> ------------------------------
>
> In backend/distributor/api.py
> <https://github.com/Entea/covid-supply-info/pull/130#discussion_r405541580>
> :
>
> > @@ -168,13 +169,26 @@ class ContactMessageAPIView(APIView):
>      @staticmethod
>      def post(request, *args, **kwargs):
>          serializer = ContactMessageSerializer(data=request.data)
> -        if serializer.is_valid():
> -            ContactMessage.objects.create(
> -                full_name=serializer.data['full_name'],
> -                phone_number=serializer.data['phone_number'],
> -                email=serializer.data['email'],
> -                title=serializer.data['title'],
> -                body=serializer.data['body'],
> -            )
> -            return Response({'success': True}, status=status.HTTP_201_CREATED)
> -        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
> +        serializer.is_valid(raise_exception=True)
>
> It already returns normal response, I just reduce code lines, nothing has
> changed, it remains as before. Instead of checking inside if statement we
> can set raise_exception=True and it will return Bad Request with status
> code 400 and response body will be with detailed error messages
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Entea/covid-supply-info/pull/130#discussion_r405541580>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AATD6EMK3JCMMOF24J2ZDFLRLR6SLANCNFSM4MBC3XJA>
> .
>

Entea committed:
Добавить возможность скрывать больницы #133 (#134)

* api.py - ограничение вывода в апи
* admin.py - возможность фильтрации списка
* models.py - новое поле
* 0039_hospital_hidden.py - миграция для нового поля

isabek committed:
#135 Add donator type government. (#136)

- Remove price_per_piece field from DonationDetail.
- Add total_cost field to DonationDetail.
- Add GOVERNMENT as a donator type.
- Fix `Blocked by X-Frame-Options Policy` issue.

Entea committed:
Merged migration paths (#137)

Entea committed:
Merged settings from prod server (#138)

Entea committed:
Фикс страницы с donations (#139)

Entea commented:
Необходимо:
* отребейзить ветку на мастер
* добавить модель DistributionDetail (скопировать DonationDetail)
* удалить Distribution#donations, связь с Donation будет через DistributionDetail#donation
* убедиться, что после ребейза можно запустить `migrate` (смерджить конфликтующие миграции, либо накатить все по новой)

Entea commented on backend/distributor/models.py:
```diff
@@ -340,3 +342,26 @@ class Meta:
+    delivered_at = models.DateField(verbose_name=_('Дата доставки'), blank=True, null=True)
+    status = models.CharField(max_length=20, choices=DISTRIBUTION_STATUSES, verbose_name=_('Статус'),
+                              default=READY_TO_SEND)
+    donations = models.ManyToManyField(Donation, verbose_name=_('Пожертвования'))
```
поле удаляем

Entea reviewed (changes requested):
В связи с уточненными требованиями, необходимо чуть-чуть допилить код

katalyzator committed:
added new model

katalyzator committed:
added new model

katalyzator committed:
Merge branch 'master' of https://github.com/Entea/covid-supply-info into issues/128_distributon

katalyzator committed:
merged migrations

katalyzator committed:
merged migrations

katalyzator committed:
Merge branch 'master' of https://github.com/Entea/covid-supply-info into issues/128_distributon

katalyzator committed:
Merge branch 'issues/128_distributon' of https://github.com/Entea/covid-supply-info into issues/128_distributon

katalyzator committed:
DistrDetail added to admin

katalyzator committed:
added detail to DistributionSerializer
